luaui/Widgets/gui_unit_market.lua

```diff
@@ -0,0 +1,480 @@
+        if team == teamID then
+            return playerID
+        end
+    end
+    return nil
+end
+
+local function getTeamName(teamID)
+    local _, _, _, isAITeam = spGetTeamInfo(teamID)
+    if isAITeam then
+        local _, _, _, aiName = spGetAIInfo(teamID)
+        if aiName then
+            local niceName = spGetGameRulesParam('ainame_' .. teamID)
+            if niceName then
+                return niceName
+            else
+                return aiName
+            end
+        else
+            return "AI Team (" .. tostring(teamID)..")"
+        end
+    else
+        local playerID = FindPlayerIDFromTeamID(teamID)
+        if playerID then
+            local playerName, _ = spGetPlayerInfo(playerID, false)
+            return playerName
+        else
+            return "Unknown Team (" .. tostring(teamID)..")"
+        end
+    end
+end
+
+local function unitSale(unitID, price, msgFromTeamID)
+    local unitDefID = spGetUnitDefID(unitID)
+    if not unitDefID then return end
+    local unitDef = UnitDefs[unitDefID]
+    if not unitDef then return end
+    local name = getTeamName(msgFromTeamID)
+    if price > 0 then
+        unitsForSale[unitID] = price
+        local msg = Spring.I18N('ui.unitMarket.sellingUnit', { name = name, unitName = unitDef.translatedHumanName, price = price })
+        Print(msg)
+    else
+        ClearUnitData(unitID)
+    end
+end
+
+local function unitSold(unitID, price, old_ownerID, msgFromTeamID)
+    ClearUnitData(unitID)
+    local unitDefID = spGetUnitDefID(unitID)
+    if not unitDefID then return end
+    local unitDef = UnitDefs[unitDefID]
+    if not unitDef then return end
+    local old_owner_name = getTeamName(old_ownerID)
+    local new_owner_name = getTeamName(msgFromTeamID)
+    if (old_owner_name and new_owner_name) then
+        local msg = Spring.I18N('ui.unitMarket.unitSold', { oldName = old_owner_name, name = new_owner_name, unitName = unitDef.translatedHumanName, price = price })
+        Print(msg)
+    end
+end
+
+local function unitSaleBroadcast(unitID, price, msgFromTeamID)
+    unitSale(unitID, price, msgFromTeamID)
+end
+
+local function unitSoldBroadcast(unitID, price, old_ownerID, msgFromTeamID)
+    unitSold(unitID, price, old_ownerID, msgFromTeamID)
+end
+
+function widget:Initialize()
+    -- if market is disabled, exit
+    if not unitMarket or unitMarket ~= true then
+        widgetHandler:RemoveWidget() -- not enabled? shutdown
+    end
+	-- TODO, in 1vs1 or if you are alone in a team, unless for debug purposes - widget should auto-shutdown
+    -- TODO, if you restart the widget you will forget who is selling any units, create inline that loops through all units and gets param if they are on sale
+    if not(Spring.IsReplay() or spGetSpectatingState()) then
+	    widgetHandler:AddAction("sell_unit", OfferToSellAction, nil, 'p')
+    end
+	widgetHandler:RegisterGlobal('unitSaleBroadcast', unitSaleBroadcast)
```

Comment:
if this is to be an api for multiple widgets then this should be handled by the widget handler

Reply:
Do you mean giving some wrapped function available to all widgets that calls `spSendLuaRulesMsg("unitOfferToSell " .. unitID)` like this one does?

Reply:
Got it. Will do in a few minutes.

Reply:
https://github.com/beyond-all-reason/Beyond-All-Reason/pull/2879/commits/431cb5334105c43d98129c9f0f148ae2788dcea9 not sure I understood everything right. But it seems to work.

Reply:
https://github.com/beyond-all-reason/Beyond-All-Reason/pull/2879/commits/2f0ff1503ed6a19ea243307e976336a515d19bb6 oh yeahm buy callin too. Tested, both work.

Reply:
no, the other way around, it was almost ok before:
* widget calls `Spring.SendLuaRulesMsg("unitOfferToSell " .. unitID)` (like it was before)
* gadget checks if the unit is valid etc, and then calls `Script.LuaUI.unitSaleBroadcast(unitID)` (like it was before)
* widget listener for this event:
```diff
- function unitSaleBroadcast(unitID) blabla end
- widgetHandler:RegisterGlobal('unitSaleBroadcast', unitSaleBroadcast)

+ function widget:unitSaleBroadcast(unitID) blabla end
```

Reply:
https://github.com/beyond-all-reason/Beyond-All-Reason/pull/2879/commits/a0471bf02bb2fd31b6aecb8fe7d695ace5756dc0 ok, changed, works.

Reply:
looks decent